nncf/experimental/torch/search_building_blocks/search_blocks.py

```diff
@@ -0,0 +1,679 @@
+        end_i_node = combination[i].end_node
+        start_nexti_node = combination[i + 1].start_node
+        if end_i_node.node_key in start_nexti_node.node_key:
+            i += 1
+            continue
+        return False
+    return True
+
+def search_lin_combination(block: BuildingBlock, blocks: List[BuildingBlock]) -> bool:
+    """
+    Checks that a given block is linear combination of some blocks.
+    A linear combination of blocks is a sequence of blocks following each other in the graph
+    and connected by one edge.
+    """
+    max_num = len(blocks)
+    all_combinations = []
+    for i in range(max_num, 1, -1):
+        all_combinations = list(combinations(blocks, i))
+        for combo in all_combinations:
+            if check_blocks_combination_is_block(block, combo):
+                return True
+    return False
+
+def remove_linear_combination(sorted_building_blocks: List[BuildingBlock]) -> List[BuildingBlock]:
+    """
+    Search and remove of block which is a combination of other blocks following each other.
+    """
+    result_blocks = []
+    start_to_idx = {}
+    last_node = None
+    for block in sorted_building_blocks:
+        if last_node == block.end_node:
+            if block.start_node.main_id in start_to_idx:
+                if not search_lin_combination(block, result_blocks[start_to_idx[block.start_node.main_id]:]):
+                    result_blocks.append(block)
+                    last_node = block.end_node
+        else:
+            result_blocks.append(block)
+            last_node = block.end_node
+            if block.start_node.main_id not in start_to_idx:
+                start_to_idx[block.start_node.main_id] = len(result_blocks) - 1
+
+    return result_blocks
+
+def restore_node_name_in_orig_graph(building_blocks: List[BuildingBlock], orig_graph: PTNNCFGraph) -> List[str]:
+    """
+    Restore the original names of the start and end of the block in original graph.
+    """
+    building_block_in_orig_format = []
+    for block in building_blocks:
+        id_st = block.start_node.bottom_id # dummy node
+        id_end = block.end_node.bottom_id
+        block_in_orig_format = [orig_graph.get_node_key_by_id(id_st).split(' ')[-1],
+                                orig_graph.get_node_key_by_id(id_end).split(' ')[-1]]
+        building_block_in_orig_format.append(block_in_orig_format)
+    return building_block_in_orig_format
+
+def get_potential_candidate_for_block(sgraph: SearchGraph) -> Tuple[Dict[str, List[int]]]:
+    act_input_shape = {} # key - str(shape), value - set of node_keys
+    act_output_shape = {} # key - str(shape), value - set of node_keys
+    for node in sgraph.get_all_nodes():
+        next_edges = sgraph.get_next_edges(node.node_key)
+        prev_edges = sgraph.get_prev_edges(node.node_key)
+        for _, edge_attr in next_edges.items():
+            sgraph.set_node_attr(node.node_key, 'activation_output_shape', edge_attr['activation_shape'])
```

Comment:
Do not use string literals such as `'activation_shape'`, use references to predefined constants.

Reply:
Done.